Lib/test/test_ctypes/test_dlerror.py

```diff
@@ -0,0 +1,90 @@
+from ctypes import CDLL
+
+FOO_C = r"""
+#include <stdio.h>
+
+/* This is a 'GNU indirect function' (IFUNC) that will be called by
+   dlsym() to resolve the symbol "foo" to an address. Typically, such
+   a function would return the address of an actual function, but it
+   can also just return NULL.  For some background on IFUNCs, see
+   https://willnewton.name/uncategorized/using-gnu-indirect-functions/
+
+   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014
+*/
+
+asm (".type foo, @gnu_indirect_function");
+
+void *foo(void)
+{
+    fprintf(stderr, "foo IFUNC called\n");
+    return NULL;
+}
+"""
+
+
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'Test only valid for Linux')
+class TestNullDlsym(unittest.TestCase):
+    """GH-126554: Ensure that we catch NULL dlsym return values
+
+    In rare cases, such as when using GNU IFUNCs, dlsym(),
+    the C function that ctypes' CDLL uses to get the address
+    of symbols, can return NULL.
+
+    The objective way of telling if an error during symbol
+    lookup happened is to call glibc's dlerror() and check
+    for a non-NULL return value.
+
+    However, there can be cases where dlsym() returns NULL
+    and dlerror() is also NULL, meaning that glibc did not
+    encounter any error.
+
+    In the case of ctypes, we subjectively treat that as
+    an error, and throw a relevant exception.
+
+    This test case ensures that we correctly enforce
+    this 'dlsym returned NULL -> throw Error' rule.
+    """
+
+    def test_null_dlsym(self):
+        import subprocess
+        import tempfile
+
+        try:
+            p = subprocess.Popen(['gcc', '--version'], stdout=subprocess.PIPE,
+                                 stderr=subprocess.DEVNULL)
+            out, _ = p.communicate()
+        except OSError:
+            raise unittest.SkipTest('gcc, needed for test, not available')
+        with tempfile.TemporaryDirectory() as d:
+            # Create a source file foo.c, that uses
+            # a GNU Indirect Function. See FOO_C.
+            srcname = os.path.join(d, 'foo.c')
+            libname = 'py_ctypes_test_null_dlsym'
+            dstname = os.path.join(d, 'lib%s.so' % libname)
+            with open(srcname, 'w') as f:
+                f.write(FOO_C)
+            self.assertTrue(os.path.exists(srcname))
+            # Compile the file to a shared library
+            cmd = ['gcc', '-fPIC', '-shared', '-o', dstname, srcname]
+            out = subprocess.check_output(cmd)
+            self.assertTrue(os.path.exists(dstname))
+            # Load the shared library
+            L = CDLL(dstname)
+
+            with self.assertRaises(AttributeError) as cm:
+                # Try accessing the 'foo' symbol.
+                # It should resolve via dlsym() to NULL,
+                # and since we subjectively treat NULL
+                # addresses as errors, we should get
+                # an error.
+                L.foo
```

Comment:
Can we assert that this is equal to something?

Reply:
The precise thing to assert here is that an exception should be raised immediately after executing this command.

Therefore, I suggest adding a `self.fail()` with a comment, similar to what we do in https://github.com/python/cpython/blob/54c63a32d06cb5f07a66245c375eac7d7efb964a/Lib/test/test_ctypes/test_structunion.py#L347

WDYT?

Reply:
I'm not sure how the `with self.assertRaises(AttributeError) as cm:` block works.

Will it break from this block if an AttributeError is raised? Is this a wrapper around `try/except`?

Reply:
I tried this and verified it works. Pushed another commit with an explicit failure if not attribute error was raised.

Reply:
I hadn't seen your comment, you can see the commit and judge if you want to keep it or not :)